common/src/main/java/com/proximyst/ban/service/MessageService.java

```diff
@@ -0,0 +1,230 @@
+import com.google.inject.Singleton;
+import com.proximyst.ban.config.MessageKey;
+import com.proximyst.ban.config.MessagesConfig;
+import com.proximyst.ban.factory.IBanExceptionalFutureLoggerFactory;
+import com.proximyst.ban.factory.IMessageFactory;
+import com.proximyst.ban.message.IMessage;
+import com.proximyst.ban.message.IMessageComponent;
+import com.proximyst.ban.model.BanUser;
+import com.proximyst.ban.model.Punishment;
+import com.proximyst.ban.platform.IBanAudience;
+import com.proximyst.ban.platform.IBanServer;
+import com.proximyst.ban.utils.BanExceptionalFutureLogger;
+import java.text.SimpleDateFormat;
+import java.util.concurrent.CompletableFuture;
+import net.kyori.adventure.identity.Identity;
+import net.kyori.adventure.text.minimessage.MiniMessage;
+import org.apache.commons.lang3.time.DurationFormatUtils;
+import org.checkerframework.checker.index.qual.NonNegative;
+import org.checkerframework.checker.nullness.qual.NonNull;
+
+// The existence of this class is because of this issue:
+// <https://github.com/google/guice/issues/1345>.
+@Singleton
+public final class MessageService {
+  private static final @NonNull String KEY_TARGET_NAME = "targetName";
+  private static final @NonNull String KEY_TARGET_UUID = "targetUuid";
+  private static final @NonNull String KEY_PUNISHER_NAME = "punisherName";
+  private static final @NonNull String KEY_PUNISHER_UUID = "punisherUuid";
+  private static final @NonNull String KEY_PUNISHMENT_ID = "punishmentId";
+  private static final @NonNull String KEY_PUNISHMENT_DATE = "punishmentDate";
+  private static final @NonNull String KEY_PUNISHMENT_REASON = "reason";
+  private static final @NonNull String KEY_PUNISHMENT_EXPIRY = "expiry";
+  private static final @NonNull String KEY_PUNISHMENT_DURATION = "duration";
+  private static final @NonNull String KEY_PUNISHMENT_TYPE = "punishmentType";
+  private static final @NonNull String KEY_PUNISHMENT_VERB = "punishmentVerb";
+  private static final @NonNull String KEY_QUANTITY = "amount";
+
+  private final @NonNull IMessageFactory messageFactory;
+  private final @NonNull MessagesConfig messagesConfig;
+  private final @NonNull IUserService userService;
+  private final @NonNull BanExceptionalFutureLogger<?> banExceptionalFutureLogger;
+  private final @NonNull IBanServer banServer;
+
+  @Inject
+  public MessageService(final @NonNull IMessageFactory messageFactory,
+      final @NonNull MessagesConfig messagesConfig,
+      final @NonNull IUserService userService,
+      final @NonNull IBanExceptionalFutureLoggerFactory banExceptionalFutureLoggerFactory,
+      final @NonNull IBanServer banServer) {
+    this.messageFactory = messageFactory;
+    this.messagesConfig = messagesConfig;
+    this.userService = userService;
+    this.banExceptionalFutureLogger = banExceptionalFutureLoggerFactory.createLogger(this.getClass());
+    this.banServer = banServer;
+  }
+
+  // <editor-fold desc="Errors" defaultstate="collapsed">
```

Comment:
editor-folds aren't cool (there are a few of them in this class)

Reply:
They are when there are many, many methods :-)